igvc_perception/src/train_eval/train.py

```diff
@@ -0,0 +1,217 @@
+                    help='random seed (default: 1)')
+parser.add_argument('--cfgfile', type=str, default='cfg/igvc.cfg',
+                    help='Directory containing cfg for train and evaluation.')
+parser.add_argument('--test', action='store_true', default=False,
+                    help='Skip training, and evaluate a loaded model on test data.')
+parser.add_argument('--val_samples', type=int, default=10,
+                    help='Number of validation samples to use from train data.')
+
+args = parser.parse_args()
+
+args.cuda = not args.no_cuda and torch.cuda.is_available()
+torch.manual_seed(args.seed)
+if args.cuda:
+    print('Using cuda.')
+    torch.cuda.manual_seed(args.seed)
+
+kwargs = {'num_workers': 1, 'pin_memory': True} if args.cuda else {}
+
+transform = transforms.Compose([
+                 transforms.ToTensor(),
+            ])
+
+cfg_params = utils.read_data_cfg(args.cfgfile)
+
+train_txt = cfg_params['train']
+test_txt = cfg_params['test']
+backup_dir = cfg_params['backup']
+
+if args.load_model is not None:
+    print('Loading model from %s.' % args.load_model)
+    model = models.model.UNet(args.im_size, args.kernel_size)
+    model.load_state_dict(torch.load(args.load_model))
+elif args.test:
+    print('Missing model file for evaluating test set.')
+    exit()
+else:
+    model = models.model.UNet(args.im_size, args.kernel_size)
+
+# Datasets and dataloaders.
+if not args.test:
+    train_dataset = IGVCDataset(train_txt, im_size=args.im_size, split='train', transform=transform, val_samples=args.val_samples)
+    val_dataset = IGVCDataset(train_txt, im_size=args.im_size, split='val', transform=transform, val_samples=args.val_samples)
+    train_loader = torch.utils.data.DataLoader(train_dataset, batch_size=args.batch_size, shuffle=True, **kwargs)
+    val_loader = torch.utils.data.DataLoader(val_dataset, batch_size=args.batch_size, shuffle=True, **kwargs)
+    
+    # Optmizer
+    lr = args.lr
+    print('Initial lr: %f.' % lr)
+    optimizer = optim.Adam(model.parameters(), lr=lr, weight_decay=args.weight_decay)
+else:
+    test_dataset = IGVCDataset(test_txt, im_size=args.im_size, split='test', transform=transform)
+    test_loader = torch.utils.data.DataLoader(test_dataset, batch_size=1, shuffle=True, **kwargs)
+
+criterion = F.binary_cross_entropy
+if args.cuda:
+    model.cuda()
+
+def train(epoch):
+    iters = []
+    lrs = []
+    train_losses = []
+    val_losses = []
+    val_accuracies = []
+    model.train()
+    # train loop
+    for batch_idx, batch in enumerate(train_loader):
+        # prepare data
+        images = Variable(batch[0])
+        targets = Variable(batch[1])
+
+        if args.cuda:
+            images, targets = images.cuda(), targets.cuda()
+
+        optimizer.zero_grad()
+        outputs = model(images)
+        loss = criterion(outputs, targets)
+        loss.backward()
+        optimizer.step()
+
+        if args.vis and batch_idx % args.log_interval == 0 and images.shape[0] == 1:
```

Comment:
I find that, if I make `batch_size` > 1, i'm no longer able to visualize the model's output

Reply:
I was able to "fix" this by removing the `images.shape[0] == 1` condition. Regardless of the number of batches, `outputs.cpu().data.numpy().shape` always has the same format: [b,n_c,w,h], where b is the number of batches:

```python
if args.vis and (batch_idx % args.log_interval == 0):

            cv2.imshow('output: ', np.hstack(tuple([outputs.cpu().data.numpy()[i][0] for i in range(outputs.cpu().data.numpy().shape[0])])))
            cv2.imshow('target: ', np.hstack(tuple([targets.cpu().data.numpy()[i][0] for i in range(outputs.cpu().data.numpy().shape[0])])))
            cv2.waitKey(10)
```

Reply:
Resolving as this has no effect on functionality